### PR TITLE
Use OpenDB as the Primary Layout Format

### LINKED ISSUE #1159: Use .odb rather than .def between OR steps
### Prompt
Between steps using openroad the intermediate files should be a .odb rather than a .def.  We are storing additional information that can't be represented in def (guides, hierarchy, etc).  ORFS has already made the switch as a reference.

### Proposal
Use read_db / write_db rather than read_def / write_def

### PR DESCRIPTION
**Pull request by donn**
## Flow Scripts
\+ Support OpenROAD ODB format as the primary format in which data is stored after placement
\+ Create `manipulate_layout` to centralize calling odbpy scripts
\+ Add `CURRENT_POWERED_NETLIST` to list of "current" variables, separate from 
`CURRENT_NETLIST` to avoid confusion
~ Update `scripts/odbpy/reader.py:OdbReader` and *all dependent scripts* now to use `odb` inputs and outputs. Add new `--output-def` flag to continue saving a DEF layout of the result.
~ Verbosity now a global variable, not an environment variable
~ Streamline saving using `scripts/utils/utils.tcl:run_openroad_script`: new `-save` option added that will automatically handle setting and unsetting `SAVE_` variables and `CURRENT_` variables
~ Rewrite Diode Insertion Strategy 3 iterations to be readable by human beings
~ `scripts/tcl_commands/all.tcl:heal_antenna_violators` now only uses `CURRENT_DEF`
~ `scripts/tcl_commands/all.tcl:set_netlist` updated to support flags and arguments properly, `-lec` now only works if `LEC_ENABLE` is set to `1`. Invocations updated
~ Usage of `scripts/tcl_commands/all.tcl:trim_lib` updated: no longer has any optional arguments. Documentation updated.
~ `scripts/tcl_commands/all.tcl:write_verilog` now uses `write_views.tcl` with the `-save` option. Documentation updated to reflect the fact it no longer calls `set_netlist`.
~ Fix various python invocations
~ Move `scripts/tcl_commands/routing.tcl:gen_pdn` to `floorplan`
~ Update `scripts/or_issue.py` to handle `odb` inputs and outputs
~ Update `scripts/tcl_commands/lvs.tcl:write_powered_verilog`, `scripts/tcl_commands/routing.tcl:power_routing` to use `-odb` as an input instead of `-def` (and, if applicable, `-output_odb` *alongside* `-output_def`)
\- Remove various scripts that rely on text manipulation (incl. `scripts/simple_cts.py`- thank you for your service)
\- Remove vestiges of site widening
\- Remove `-canonical` optional from `scripts/tcl_commands/all.tcl:write_verilog` (dubious/undocumented utility)
\- Remove `verilog_to_verilogPower` (unused)
\- Removed `zeroize_origin_lef`
\- Remove `PREV_NETLIST` variable
\- Remove various invocations of `scripts/tcl_commands/utils.tcl:write_verilog`, replace with `-save` arguments

## OpenROAD Scripts
\+ Create OpenROAD script `common` folder, isolating common code into a folder instead of the preceding chaos
\+ OpenROAD scripts now use a unified script for reading and writing views: `scripts/openroad/common/io.tcl`: `read` reads the `.odb` database + relevant liberty and sdc files and `write` writes any views specified as `SAVE_` variables: see the file for documentation
~ Update Copyright & Licensing for all OpenROAD scripts
~ Rewrite `scripts/odbpy/defutil.py:merge_components` to use odb
~ `scripts/odbpy/defutil.py:replace_pins` -> `relocate_pins`, with a partial rewrite and re-documentation.
~ `scripts/odbpy/defutil.py:add_def_obstructions` -> `add_obstructions`
~ `scripts/openroad/write_verilog.tcl` -> `write_view.tcl`, since it is literally just a "read" and a "write" call now
~ Partially rewrite `scripts/odbpy/power_utils.py:write_powered_def`, the most disgusting function in the codebase
\- Remove `scripts/odbpy/defutil.py:merge_pins` (unused)

---
Resolves #1159

### CODE REVIEW

**Comment by donn** [comment]
@maliberty I would like for you to give a cursory review of just the OpenROAD scripts. See if anything jumps out to you as obviously wrong/missing.

**Comment by donn** [comment]
Blocked on https://github.com/The-OpenROAD-Project/OpenROAD/issues/2244

**Comment by maliberty** [comment]
I prefer smaller more focused PRs as they are easier to review.  I don't see any obvious problems but I haven't closely read through it all.

**Comment by donn** [comment]
That's good enough for me, all I need is a high-level sanity check.

File: docs/source/openlane_commands.md [hunk 1]
```diff
@@ -18,12 +18,12 @@ Most of the following commands' implementation exists in this [file][0]
 | Command      | Flags                   | Description                                           |
 |---------------|------------------------|-----------------------------------------|
 | `set_netlist <netlist>`   | | Sets the current netlist used by the flow to `<netlist>` |
-|    | `[-lec]` | Runs logic verification for the new netlist against the previous netlist. <br> Optional flag.       |
+|    | `[-lec]` | Runs logic verification for the new netlist against the previous netlist, if `LEC_ENABLE` is set to 1. <br> Optional flag. |
```
Comment (kareefardi) [inline docs/source/openlane_commands.md:21]: isn't the whole purpose of using `-lec` is to force it despite the env variable?

File: scripts/odbpy/defutil.py [hunk 2]
```diff
@@ -52,249 +51,92 @@ def extract_core_dims(output_data, input_lef, input_def):
...
+    Adds all components in a donor DEF file that do not exist in the (recipient) INPUT_DEF.
+
+    Existing components with the same name will *not* be overwritten.
     """
-    move_diearea(input_lef, output_def, template_def)
+    donor = OdbReader(input_lef, donor_def)
+    recipient = reader
+
+    for instance in donor.instances:
+        odb.dbInst_create(recipient.block, instance.getMaster(), instance.getName())
```
Comment (kareefardi) [inline scripts/odbpy/defutil.py:83]: will this also place the instance to the same position in the donor def?

File: scripts/odbpy/power_utils.py [hunk 3]
```diff
@@ -78,41 +81,39 @@ def find_power_ground_port(port_name, ports):
...
+
+    nets_not_found = False
+    if vdd_net is None:
+        print(f"Power port {power_port} not found in design.", file=sys.stderr)
+        nets_not_found = True
+    if gnd_net is None:
+        print(f"Ground port {ground_port} not found in design.", file=sys.stderr)
+        nets_not_found = True
+    if nets_not_found:
+        exit(0)
```
Comment (kareefardi) [inline scripts/odbpy/power_utils.py:110]: that's a zero exit for ,i suppose, an error

File: scripts/odbpy/power_utils.py [hunk 4]
```diff
@@ -138,149 +139,123 @@ def find_power_ground_port(port_name, ports):
...
-            for port in power.block.getBTerms()
-            if port.getName() in POWER_GROUND_PORT_NAMES
+        pg_port_names = [port.getName() for port in VDD_PORTS + GND_PORTS]
+        pg_ports = [
+            port for port in power.block.getBTerms() if port.getName() in pg_port_names
         ]
 
-        for port in power_ground_ports:
-            iterms = port.getNet().getITerms()
-            for iterm in iterms:
```
Comment (kareefardi) [inline scripts/odbpy/power_utils.py:257]: this loop got (accidentally?) removed

File: scripts/openroad/common/io.tcl [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+        flags {-multi_corner}
+
+    set libs $::env(LIB_SYNTH_COMPLETE)
+
+    if { [info exists keys(-override)] } {
+        set libs $::env(LIB_SYNTH_COMPLETE)
+    }
+
+    if { [info exists flags(-multi_corner)] } {
+        define_corners ss tt ff; # tt unintuitively first because that makes it the "default"
```
Comment (kareefardi) [inline scripts/openroad/common/io.tcl:48]: tt is not first

File: scripts/openroad/common/io.tcl [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+
+proc read_libs {args} {
+    sta::parse_key_args "read_libs" args \
+        keys {-override}\
+        flags {-multi_corner}
+
+    set libs $::env(LIB_SYNTH_COMPLETE)
+
+    if { [info exists keys(-override)] } {
+        set libs $::env(LIB_SYNTH_COMPLETE)
```
Comment (kareefardi) [inline scripts/openroad/common/io.tcl:44]: shouldn't this be `set libs $keys(-override)`

File: scripts/openroad/ioplacer.tcl [hunk 7]
```diff
@@ -11,35 +11,26 @@
...
-	puts stderr $errmsg
-	exit 1
-}
-
-if {[catch {read_def $::env(CURRENT_DEF)} errmsg]} {
-	puts stderr $errmsg
-	exit 1
 }
 
 if {$::env(FP_IO_HLENGTH) != "" && $::env(FP_IO_HLENGTH) != ""} {
```
Comment (kareefardi) [inline scripts/openroad/ioplacer.tcl:21]: not part of this pr but ummm

File: scripts/openroad/rcx.tcl [hunk 8]
```diff
@@ -11,28 +11,9 @@
 # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 # See the License for the specific language governing permissions and
 # limitations under the License.
+source $::env(SCRIPTS_DIR)/openroad/common/io.tcl
+read
```
Comment (kareefardi) [inline scripts/openroad/rcx.tcl:15]: looks like it was reading `$::env(RCX_LIB)` before

Comment (kareefardi) [reply scripts/openroad/rcx.tcl:15]: it might be the case that there is a different odb using `RCX_LIB` and `RCX_LEF` but I didn't get to that yet.

File: scripts/openroad/sta.tcl [hunk 9]
```diff
@@ -11,37 +11,16 @@
...
+        read
     } else {
-        if {[catch {read_verilog $::env(CURRENT_NETLIST)} errmsg]} {
-            puts stderr $errmsg
-            exit 1
-        }
-        link_design $::env(DESIGN_NAME)
+        read_libs
+        read_lef $::env(MERGED_LEF)
+        read_netlist
```
Comment (kareefardi) [inline scripts/openroad/sta.tcl:21]: `read_sdc` too

Comment (kareefardi) [reply scripts/openroad/sta.tcl:21]: is `link_design` not needed too ?

File: scripts/openroad/common/io.tcl [hunk 10]
```diff
@@ -0,0 +1,163 @@
...
+    }
+
+    if { [info exists ::env(SAVE_GUIDE)] } {
+        puts "Writing routing guides to $::env(SAVE_GUIDE)..."
+        write_guides $::env(SAVE_GUIDE)
+    }
+
+    if { [info exists ::env(SAVE_SDF)] } {
+        puts "Writing SDF to $::env(SAVE_SDF)..."
+        write_sdf $::env(SAVE_SDF)
```
Comment (kareefardi) [inline scripts/openroad/common/io.tcl:161]: we were always using `-include_typ` and `-divider .`

File: scripts/odbpy/defutil.py (hunk 2 above)
Comment (donn) [reply scripts/odbpy/defutil.py:83]: No, that's not the intention.

File: scripts/openroad/common/io.tcl (hunk 5 above)
Comment (donn) [reply scripts/openroad/common/io.tcl:48]: Oh, right, forgot to remove the comment. I tried to make all scripts read all three corners. Needless to say, didn't go well.

File: scripts/openroad/common/io.tcl (hunk 6 above)
Comment (donn) [reply scripts/openroad/common/io.tcl:44]: ... yes

File: scripts/openroad/ioplacer.tcl (hunk 7 above)
Comment (kareefardi) [reply scripts/openroad/ioplacer.tcl:21]: I meant to comment on this `if {$::env(FP_IO_HLENGTH) != "" && $::env(FP_IO_HLENGTH) != ""} {`

File: scripts/utils/utils.tcl [hunk 11]
```diff
@@ -496,5 +406,257 @@ proc cat {args} {
...
+        {-output_def optional}
+    }
+
+    set flags {}
+
+    parse_key_args "manipulate_layout" args arg_values $options flag_map $flags
+
+    set_if_unset arg_values(-indexed_log) /dev/null
+    set_if_unset arg_values(-input) $::env(CURRENT_ODB)
+    set_if_unset arg_values(-output) $arg_values(-input)
```
Comment (kareefardi) [inline scripts/utils/utils.tcl:424]: so the default output is inplace change of the odb file?

File: scripts/openroad/ioplacer.tcl (hunk 7 above)
Comment (donn) [reply scripts/openroad/ioplacer.tcl:21]: You know whose avatar this is. :) 
image

File: scripts/openroad/sta.tcl (hunk 9 above)
Comment (donn) [reply scripts/openroad/sta.tcl:21]: Both are part of `read_netlist`

File: scripts/utils/utils.tcl (hunk 11 above)
Comment (donn) [reply scripts/utils/utils.tcl:424]: yep

File: scripts/odbpy/power_utils.py (hunk 3 above)
Comment (donn) [reply scripts/odbpy/power_utils.py:110]: You are right

File: scripts/odbpy/power_utils.py [hunk 12]
```diff
@@ -138,149 +139,126 @@ def find_power_ground_port(port_name, ports):
...
 
-        openroad_script = "\n".join(openroad_script)
-
-        openroad_stdout, openroad_stderr = p.communicate(openroad_script)
-        print(f"STDOUT: {openroad_stdout.strip()}")
-        print(f"STDERR: {openroad_stderr.strip()}")
-        print("openroad exit code:", p.returncode)
-        assert p.returncode == 0, p.returncode
-        assert os.path.exists(tmp_def_file), "DEF file doesn't exist"
+        subprocess.run(["openroad"], check=True, input=openroad_script.encode("utf8"))
```
Comment (kareefardi) [inline scripts/odbpy/power_utils.py:218]: subprocess.run won't exit the main script if the process fails right?

File: scripts/tcl_commands/all.tcl [hunk 13]
```diff
@@ -107,10 +120,11 @@ proc prep_lefs {args} {
         if { [info exists ::env(METAL_LAYER_NAMES)] } {
             set ::env(TECH_METAL_LAYERS) $::env(METAL_LAYER_NAMES)
         } else {
-            try_catch openroad -python\
+            try_catch $::env(OPENROAD_BIN) -python\
```
Comment (kareefardi) [inline scripts/tcl_commands/all.tcl:123]: should it be like this or the other way around? i think i have seen changes that do the opposite

File: scripts/tcl_commands/lvs.tcl [hunk 14]
```diff
@@ -93,20 +99,16 @@ proc run_lvs {{layout "$::env(EXT_NETLIST)"}} {
...
-            -log $::env(signoff_logs)/write_verilog.log\
-            -def_log $::env(signoff_logs)/write_powered_def.log
-
-        set_netlist $powered_netlist_name
-
-        if { $::env(LEC_ENABLE) } {
-            logic_equiv_check -rhs $::env(PREV_NETLIST) -lhs $::env(CURRENT_NETLIST)
-        }
+            -log $::env(signoff_logs)/write_powered_nl.log\
+            -def_log $::env(signoff_logs)/write_powered_df.log
```
Comment (kareefardi) [inline scripts/tcl_commands/lvs.tcl:111]: `write_powered_df.log`

File: scripts/tcl_commands/routing.tcl [hunk 15]
```diff
@@ -292,45 +254,57 @@ proc ins_diode_cells_4 {args} {
...
+
+    manipulate_layout $::env(SCRIPTS_DIR)/odbpy/defutil.py add_obstructions\
+        -indexed_log $log\
+        -output $save_db\
+        -output_def $save_def\
+        --obstructions $::env(GRT_OBS)
+
+    puts_verbose "Obstructions added over $::env(GRT_OBS)."
+
+    set_def $save_def
```
Comment (kareefardi) [inline scripts/tcl_commands/routing.tcl:303]: not that this is an issue, but for the sake of better understanding set_def isn't really neded here right?. it should have been handled in `manipulate_layout` just above it?

File: scripts/tcl_commands/routing.tcl [hunk 16]
```diff
@@ -423,31 +410,16 @@ proc run_routing {args} {
...
+        [index_file $::env(routing_tmpfiles)/grt.nl.v]\
+        -powered_to [index_file $::env(routing_tmpfiles)/grt.pnl.v]\
+        -log $::env(routing_logs)/write_verilog_global.log
 
     # detailed routing
     detailed_routing
 
-    set detailed_routed_netlist [index_file $::env(routing_tmpfiles)/detailed.v]
-
-    write_verilog $detailed_routed_netlist -log $::env(routing_logs)/write_verilog_detailed.log
```
Comment (kareefardi) [inline scripts/tcl_commands/routing.tcl:438]: was this accidentally removed or is it part of somewhere else or perhaps not needed ?

File: scripts/tcl_commands/routing.tcl [hunk 17]
```diff
@@ -459,23 +431,13 @@ proc run_resizer_timing_routing {args} {
...
-        set_def $::env(SAVE_DEF)
-        set ::env(CURRENT_SDC) $::env(SAVE_SDC)
+        run_openroad_script $::env(SCRIPTS_DIR)/openroad/resizer_routing_timing.tcl\
+            -indexed_log [index_file $::env(routing_logs)/resizer.log]\
+            -save "dir=$::env(routing_tmpfiles),def,sdc,odb,netlist,powered_netlist"
 
         TIMER::timer_stop
         exec echo "[TIMER::get_runtime]" | python3 $::env(SCRIPTS_DIR)/write_runtime.py "resizer timing optimizations - openroad"
 
-        write_verilog $::env(routing_results)/$::env(DESIGN_NAME).resized.v -log $::env(routing_logs)/write_verilog.log
```
Comment (kareefardi) [inline scripts/tcl_commands/routing.tcl:472]: same comment as the previous removal

File: scripts/tcl_commands/sta.tcl [hunk 18]
```diff
@@ -85,38 +79,32 @@ proc run_parasitics_sta {args} {
...
+                    -log $::env(signoff_logs)/rcx_sta.log\
                     -process_corner $process_corner
 
                 set ::env(LAST_TIMING_REPORT_TAG) [index_file $::env(signoff_reports)/rcx_sta]
 
                 set ::env(CURRENT_SDF) $::env(SAVE_SDF)
+                unset ::env(SAVE_SDF)
             }
-
-            run_sta\
```
Comment (kareefardi) [inline scripts/tcl_commands/sta.tcl:115]: any reason why this got removed ?

File: scripts/odbpy/power_utils.py (hunk 12 above)
Comment (donn) [reply scripts/odbpy/power_utils.py:218]: With `check=True`, it will!

File: scripts/tcl_commands/lvs.tcl (hunk 14 above)
Comment (donn) [reply scripts/tcl_commands/lvs.tcl:111]: yes?

File: scripts/tcl_commands/all.tcl (hunk 13 above)
Comment (donn) [reply scripts/tcl_commands/all.tcl:123]: like this, definitely not the other way around: either `$::env(OPENROAD_BIN) -python` or `python3` for things that don't need OpenDB.

File: scripts/tcl_commands/sta.tcl [hunk 19]
```diff
@@ -85,38 +79,32 @@ proc run_parasitics_sta {args} {
         nom MERGED_LEF RCX_RULES
     } {
         if { [info exists ::env($lef)] } {
-            set ::env(SAVE_SPEF) "$::env(SPEF_PREFIX).$process_corner.spef"
-
             run_spef_extraction\
                 -log $::env(signoff_logs)/parasitics_extraction.$process_corner.log\
                 -rcx_lib $::env(LIB_SYNTH_COMPLETE)\
                 -rcx_rules $::env($ruleset)\
-                -rcx_lef $::env($lef)\
```
[the remark this answers is not in the document]
Comment (donn) [reply scripts/tcl_commands/sta.tcl:94]: you're right- thanks for catching this

File: scripts/tcl_commands/lvs.tcl (hunk 14 above)
Comment (kareefardi) [reply scripts/tcl_commands/lvs.tcl:111]: sorry, small typo `write_powered_def.log`

Comment (donn) [reply scripts/tcl_commands/lvs.tcl:111]: does it matter

File: scripts/tcl_commands/routing.tcl (hunk 17 above)
Comment (donn) [reply scripts/tcl_commands/routing.tcl:472]: You'll find the `-save` argument a couple lines above now has : `netlist,powered_netlist`.

File: scripts/tcl_commands/routing.tcl (hunk 15 above)
Comment (donn) [reply scripts/tcl_commands/routing.tcl:303]: No- manipulate_layout doesn't call `set_def` or `set_db`.

File: scripts/tcl_commands/sta.tcl (hunk 18 above)
Comment (donn) [reply scripts/tcl_commands/sta.tcl:115]: I moved it to line 99

File: scripts/tcl_commands/lvs.tcl (hunk 14 above)
Comment (donn) [reply scripts/tcl_commands/lvs.tcl:111]: ill fix it anyway

File: scripts/tcl_commands/routing.tcl [hunk 20]
```diff
@@ -30,63 +30,65 @@ proc groute_antenna_extract {args} {
...
-            set ::env(SAVE_DEF) [index_file $::env(routing_tmpfiles)/global_$iter.def]
-            set ::env(SAVE_GUIDE) [index_file $::env(routing_tmpfiles)/global_$iter.guide]
-            set saveLOG [index_file $::env(routing_logs)/global_$iter.log]
-            set replaceWith "INSDIODE$iter"
-            puts_info "Starting antenna repair iteration $iter with $prevAntennaVal violations..."
-
-            try_catch $::env(OPENROAD_BIN) -python $::env(SCRIPTS_DIR)/odbpy/defutil.py replace_instance_prefixes\
-                --output $::env(CURRENT_DEF)\
+        puts_info "Starting OpenROAD Antenna Repair Iterations..."
+        set iter 1
```
[the remark this answers is not in the document]
Comment (donn) [reply scripts/tcl_commands/routing.tcl:49]: Added increment

File: scripts/tcl_commands/sta.tcl (hunk 19 above)
[the remark this answers is not in the document]
Comment (kareefardi) [reply scripts/tcl_commands/sta.tcl:94]: still not added ?

[the remark this answers is not in the document]
Comment (donn) [reply scripts/tcl_commands/sta.tcl:94]: Added in 50d72c96985c2a21ca05e90b3ee3aaa2ec7cc577- idk why it's not showing up in the diff

**Review by kareefardi — APPROVED** [review]
